src/compile/data/dataflow.ts

```diff
@@ -127,3 +135,23 @@ export interface DependentNode {
    */
   dependentFields: () => StringSet;
 }
+
+
+/**
+ * Trait for nodes that can be cloned.
+ */
+export interface ClonableNode {
+  /**
+   * Clone this node with a deep copy.
+   */
+  clone: () => this;
+}
+
+
+export function isNewFieldNode(node: any): node is NewFieldNode {
```

Comment:
Can't we just make produces return `null` for nodes that don't produce new field?